File: 2017/day17/p2v1.js
const test = false; // true;
const step = test ? 3 : 359;

const ub = test ? 2017 : 50000000;

const arr = new Array(ub+1);

for (let i = 1; i <= ub; i++) { arr[i] = i - 1;}
arr[0] = ub;

let t, y, x, k;

y = 4;
x = 3;
k = ub;

while (true) {
// 一つ手前y、見つけた位置xに、値kを入れる
t = arr[x];
arr[x] = -k;

if (k == 0) { break; } // 終了

arr[y] = t; // k==0のときx==yなので後にしないとおかしくなる

// tからステップ数だけ手前に戻る
x = t;
for (let i = step; i > 0; i--) {
    y = x;
    x = arr[x];
}

// k を減らす
k--;
}

if (test) {
// 2017の前後を表示
for (let i = 0; i <= 6; i++) {
    console.log(i, - +arr[i]);
}
}

// 0はxの位置に書き込んだ、その次を表示
console.log(x, +arr[x]);
const x1 = x == ub ? 0 : x + 1;
console.log(x1, - +arr[x1]);

/*
>node p2v1.js
3665345 -0
3665346 39479736
*/

File: 2021/day22/p2v1.hs
import Data.List.Split
import Data.Array
import qualified Data.IntMap as IM
import Data.List
import qualified Data.IntSet as IS
import System.CPUTime

compute2 xs = (xn*yn*zn, ans)
  where
    ls = map parseline $ lines xs
    (xls:xus:yls:yus:zls:zus:_) = transpose $ map snd ls
    (p2iX,i2wX) = partLines xls xus
    (p2iY,i2wY) = partLines yls yus
    (p2iZ,i2wZ) = partLines zls zus
    (_,xn) = bounds i2wX
    (_,yn) = bounds i2wY
    (_,zn) = bounds i2wZ
    arr = accumArray (\_ x -> x) False ((1,1,1),(xn,yn,zn))
          [ ((xi,yi,zi), b)
          | (b,[xl,xu,yl,yu,zl,zu]) <- ls
          , xi <- [p2iX IM.! xl..pred (p2iX IM.! succ xu)]
          , yi <- [p2iY IM.! yl..pred (p2iY IM.! succ yu)]
          , zi <- [p2iZ IM.! zl..pred (p2iZ IM.! succ zu)] ]
    ans = sum $ map (boxsize . fst) $ filter snd $ assocs arr
    boxsize (xi,yi,zi) = (i2wX ! xi) * (i2wY ! yi) * (i2wZ ! zi)

partLines :: [Int] -> [Int] -> (IM.IntMap Int, Array Int Int)
partLines xls xus = (pos2ind,ind2wid)
  where
    xs = IS.elems $ IS.fromList (xls ++ map succ xus)
    pos2ind = IM.fromList $ zip xs [1..]
    ind2wid = listArray (1,pred $ length xs) $ zipWith (-) (tail xs) xs

parseline :: String -> (Bool,[Int])
parseline l = (onoff == "on", map read xyzlu)
  where
    onoff:xyz:_ = words l
    xyzlu = concatMap (splitOn ".." . drop 2) $ wordsBy (',' ==) xyz

test = readFile "sample3.txt" >>= print . compute2
run2 = readFile "input.txt" >>= print . compute2

main = do
  t0 <- getCPUTime
  run2
  t1 <- getCPUTime
  print $ t1 - t0

{-
*Main> test
(1587808,2758514936282235)
*Main> run2
(570405888,

う、実装は正しくできたっぽいのに、本番が完了しない。
420行使って、毎回違うところで切り続けたら、この方式で区切りの個数はいくつできるだろうか。

>ghc -O2 p2v1.hs
[1 of 1] Compiling Main             ( p2v1.hs, p2v1.o )
Linking p2v1.exe ...
>p2v1
(570405888,1233304599156793)
55171875000000

55秒。
-}
